cpp/open3d/data/ExtractZIP.cpp

```diff
@@ -0,0 +1,251 @@
+            }
+        }
+
+        if (fout != nullptr) {
+            utility::LogDebug(" Extracting: {}", write_filename);
+
+            do {
+                err = unzReadCurrentFile(uf, buf, size_buf);
+                if (err < 0) {
+                    utility::LogWarning(
+                            "Error {} with zipfile in unzReadCurrentFile.",
+                            err);
+                    break;
+                }
+                if (err > 0)
+                    if (fwrite(buf, err, 1, fout) != 1) {
+                        utility::LogWarning("Error in writing extracted file.");
+                        err = UNZ_ERRNO;
+                        break;
+                    }
+            } while (err > 0);
+
+            if (fout) {
+                fclose(fout);
+            }
+        }
+
+        if (err == UNZ_OK) {
+            err = unzCloseCurrentFile(uf);
+            if (err != UNZ_OK) {
+                utility::LogWarning(
+                        "Error {} with zipfile in unzCloseCurrentFile.", err);
+            }
+        } else {
+            unzCloseCurrentFile(uf);
+        }
+    }
+
+    free(buf);
+    return err;
+}
+
+static void ExtractAll(unzFile uf, const std::string &password) {
+    uLong i;
+    unz_global_info64 gi;
+    int err;
+
+    err = unzGetGlobalInfo64(uf, &gi);
+    if (err != UNZ_OK) {
+        utility::LogError(
+                "Extraction failed in unzGetGlobalInfo with error code: {}.",
+                err);
+    }
+
+    for (i = 0; i < gi.number_entry; i++) {
+        err = ExtractCurrentFile(uf, password);
+        if (err != UNZ_OK) {
+            utility::LogError(
+                    "Extraction failed in ExtractCurrentFile with error code: "
+                    "{}.",
+                    err);
+        }
+
+        if ((i + 1) < gi.number_entry) {
+            err = unzGoToNextFile(uf);
+            if (err != UNZ_OK) {
+                utility::LogError(
+                        "Extraction failed in unzGoToNextFile with error code: "
+                        "{}.",
+                        err);
+            }
+        }
+    }
+}
+
+void ExtractFromZIP(const std::string &file_path,
+                    const std::string &extract_dir) {
+    unzFile uf = nullptr;
+
+    if (!file_path.empty()) {
+        char file_path_try[MAXFILENAME + 16] = "";
+
+        strncpy(file_path_try, file_path.c_str(), MAXFILENAME - 1);
+        // strncpy doesnt append the trailing nullptr, of the string is too
+        // long.
+        file_path_try[MAXFILENAME] = '\0';
+
+        uf = unzOpen64(file_path.c_str());
+        if (uf == nullptr) {
+            strcat(file_path_try, ".zip");
+            uf = unzOpen64(file_path_try);
+        }
+    }
+
+    if (uf == nullptr) {
+        utility::LogError("Failed to open file {}.", file_path);
+    }
+
+    // Change working directory to the extraction directory.
+    if (chdir(extract_dir.c_str())) {
```

Comment:
Issue: The working directory is getting changed here. So, all the unit-test data is getting dumped in this folder.
However, as we delete this directory in the cleanup step, the unit tests are failing as the functions are unable to find anything at this location. 

Need to get the current working directory before this function, and revert back to it after execution of this function.

Reply:
Fixed.

Reply:
Can `ChangeWorkingDirectory()` have potential side effects?

E.g. if there is an exception occurred in between the two `ChangeWorkingDirectory()`, the program may exit and we have only executed the first `ChangeWorkingDirectory()`. 

Is it possible to achieve extraction, without changing the directory? e.g. by prefixing the path in the actual extraction function?

Reply:
Thanks for pointing this out! Also, another concern was, that throwing exception may not ensure if the file was closed before exiting. 

Current Solution:
The easiest solution was to  `switch working dir, and close file before throwing error` in the `ExtractAll` function, as `ExtractCurrentFile` function is not expected to throw an error in the ideal case. 

It might still be an improvement if `prefixing the path` in `ExtractCurrentFile` works without changing dir. I will try that.

Reply:
Done. `changing working dir` is avoided. Also further cleaned up the functions.